c/rc_lock.c

```diff
@@ -0,0 +1,430 @@
+
+  // witness or witness lock can be empty if owner lock without rc is used
+  if (!witness_existing) return err;
+
+  BytesOptType mol_lock = witness_args.t->lock(&witness_args);
+  if (mol_lock.t->is_none(&mol_lock)) return err;
+
+  mol2_cursor_t mol_lock_bytes = mol_lock.t->unwrap(&mol_lock);
+  // convert Bytes to RcLockWitnessLock
+  RcLockWitnessLockType mol_witness_lock =
+      make_RcLockWitnessLock(&mol_lock_bytes);
+  RcIdentityOptType rc_identity_opt =
+      mol_witness_lock.t->rc_identity(&mol_witness_lock);
+  witness_lock->has_rc_identity = rc_identity_opt.t->is_some(&rc_identity_opt);
+  if (witness_lock->has_rc_identity) {
+    RcIdentityType rc_identity = rc_identity_opt.t->unwrap(&rc_identity_opt);
+    mol2_cursor_t id_cur = rc_identity.t->identity(&rc_identity);
+
+    uint8_t buff[CKB_IDENTITY_LEN] = {0};
+    uint32_t read_len = mol2_read_at(&id_cur, buff, sizeof(buff));
+    CHECK2(read_len == CKB_IDENTITY_LEN, ERROR_INVALID_MOL_FORMAT);
+    witness_lock->id.flags = buff[0];
+    memcpy(witness_lock->id.id, buff + 1, CKB_IDENTITY_LEN - 1);
+
+    witness_lock->proofs = rc_identity.t->proofs(&rc_identity);
+  }
+
+  BytesOptType signature_opt = mol_witness_lock.t->signature(&mol_witness_lock);
+  witness_lock->has_signature = signature_opt.t->is_some(&signature_opt);
+  if (witness_lock->has_signature) {
+    mol2_cursor_t signature_cursor = signature_opt.t->unwrap(&signature_opt);
+    witness_lock->signature_size = mol2_read_at(
+        &signature_cursor, witness_lock->signature, signature_cursor.size);
+    CHECK2(signature_cursor.size == witness_lock->signature_size,
+           ERROR_INVALID_MOL_FORMAT);
+  }
+  BytesOptType preimage_opt = mol_witness_lock.t->preimage(&mol_witness_lock);
+  if (preimage_opt.t->is_some(&preimage_opt)) {
+    mol2_cursor_t preimage_cursor = preimage_opt.t->unwrap(&preimage_opt);
+    witness_lock->preimage_size = mol2_read_at(
+        &preimage_cursor, witness_lock->preimage, preimage_cursor.size);
+    CHECK2(preimage_cursor.size == witness_lock->preimage_size,
+           ERROR_INVALID_MOL_FORMAT);
+  } else {
+    witness_lock->preimage_size = 0;
+  }
+
+exit:
+  return err;
+}
+
+#ifdef CKB_USE_SIM
+int simulator_main() {
+#else
+int main() {
+#endif
+  // don't move code_buff into global variable. It doesn't work.
+  // it's a ckb-vm bug: the global variable will be freezed:
+  // https://github.com/nervosnetwork/ckb-vm/blob/d43f58d6bf8cc6210721fdcdb6e5ecba513ade0c/src/machine/elf_adaptor.rs#L28-L32
+  // The code can't be loaded into frozen memory.
+  uint8_t code_buff[MAX_CODE_SIZE] __attribute__((aligned(RISCV_PGSIZE)));
+
+  int err = 0;
+
+  WitnessLockType witness_lock = {0};
+  ArgsType args = {0};
+  // this identity can be either from witness lock (witness_lock.id) or script
+  // args (args.id)
+  CkbIdentityType identity = {0};
+
+  err = parse_witness_lock(&witness_lock);
+  CHECK(err);
+
+  err = parse_args(&args);
+  CHECK(err);
+
+  if (args.has_rc_root) {
```

Comment:
When in `"administrator mode"` if witness has rc identity we should use identity from witness, if not we should use identity from `args.id` rather than return `ERROR_INVALID_MOL_FORMAT` error code.

Reply:
fixed